gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -312,34 +312,88 @@ for _, ply in pairs(player.GetHumans()) do
     DrpCanHear[ply] = {}
 end
 
--- proxy function to take load from PlayerCanHearPlayersVoice, which is called a quadratic amount of times per tick,
--- causing a lagfest when there are many players
-local function calcPlyCanHearPlayerVoice(listener)
-    if not IsValid(listener) then return end
-    DrpCanHear[listener] = DrpCanHear[listener] or {}
-    if listener:IsBot() then return end
-    local shootPos = listener:GetShootPos()
-    for _, talker in ipairs(player.GetHumans()) do
-        local talkerShootPos = talker:GetShootPos()
-        DrpCanHear[listener][talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
-            (shootPos:DistToSqr(talkerShootPos) < voiceDistance and -- voiceradius is on and the two are within hearing distance
-                (not dynv or IsInRoom(shootPos, talkerShootPos, talker))) -- Dynamic voice is on and players are in the same room
-    end
-end
-hook.Add("PlayerInitialSpawn", "DarkRPCanHearVoice", function(ply)
-    calcPlyCanHearPlayerVoice(ply)
-    if ply:IsBot() then return end
-    timer.Create(ply:UserID() .. "DarkRPCanHearPlayersVoice", 0.5, 0, fn.Curry(calcPlyCanHearPlayerVoice, 2)(ply))
-end)
+local gridSize = GM.Config.voiceDistance*2 -- Grid cell size is equal to the size of the diamater of player talking
+local floor = math.floor -- Caching floor as we will need to use it a lot
+
+-- Grid based position check
+timer.Create("DarkRPCanHearPlayersVoice", 1, 0, function()
+    DrpCanHear = {}
+    local players = player.GetHumans()
+
+    -- Create Grid
+    local grid = {}
+    local plyToGrid = {
+        {},
+        {}
+    }
+
+    -- Get the grid position of every player O(N)
+    for _, ply in ipairs(players) do
+        local pos = ply:GetShootPos()
+        local x = floor(pos.x/gridSize)
+        local y = floor(pos.y/gridSize)
+        grid[x] = grid[x] or {}
+        grid[x][y] = grid[x][y] or {}
+        grid[x][y][#grid[x][y] + 1] = ply
```

Comment:
Only if you specify a position. Otherwise it just inserts at the end, which is O(1). Though I suppose calling the function and having it do the check is more expensive than manually adding to the table.

Reply:
LuaJIT inlines it if you don't specify a position. It's faster than actually adding it manually unless you already have the length cached.